File: scripts/confluence_preproc.lua
#!/usr/bin/env lua
-- confluence_preproc.lua — Pre-process Confluence storage-format HTML into
-- standard HTML suitable for conversion to CommonMark via pandoc.
--
-- Usage (drop-in replacement for confluence_preproc.py):
--   cat storage.html | lua confluence_preproc.lua | pandoc --from=html --to=commonmark
--
-- Called automatically by confluence_publish.sh --pull.
--
-- Transformations applied (in order, matching confluence_preproc.py behaviour):
--   plantuml_src_b64 comment + <img>         → <pre><code class="language-plantuml">
--   ac:structured-macro name="code"          → <pre><code class="language-X">
--   ac:structured-macro name="info|note|.."  → <blockquote>
--   ac:structured-macro name="expand"        → unwrapped rich-text body
--   Remaining ac:structured-macro elements   → stripped (content discarded)
--   CDATA sections                           → HTML-escaped inline text
--   ac:* and ri:* namespace tags             → stripped (inner text kept)

-- ── HTML helpers ──────────────────────────────────────────────────────────────

local function html_escape(s)
  return (s:gsub('&', '&amp;')
            :gsub('<', '&lt;')
            :gsub('>', '&gt;')
            :gsub('"', '&quot;'))
end

-- Decode a base64 string using the system `base64` command (via temp file).
local function b64_decode(b64)
  local tmp = os.tmpname()
  local f = io.open(tmp, 'w')
  if not f then return nil end
  f:write(b64)
  f:close()
  local handle = io.popen('base64 -d < ' .. tmp .. ' 2>/dev/null')
  local result = handle and handle:read('*a') or ''
  if handle then handle:close() end
  os.remove(tmp)
  return result ~= '' and result or nil
end

-- Trim leading/trailing whitespace.
local function trim(s)
  return s:match('^%s*(.-)%s*$')
end

-- ── Macro converters ──────────────────────────────────────────────────────────

-- Convert a Confluence code macro block to an HTML <pre><code> block.
local function convert_code_macro(block)
  local lang = block:match('ac:name="language"[^>]*>([^<]*)</ac:parameter>') or ''
  -- CDATA body takes priority; fall back to plain-text-body
  local body = block:match('<!%[CDATA%[([%s%S]-)%]%]>')
            or block:match('<ac:plain%-text%-body>([%s%S]-)</ac:plain%-text%-body>')
            or ''
  return '<pre><code class="language-' .. trim(lang) .. '">'
      .. html_escape(trim(body))
      .. '</code></pre>'
end

-- Convert info/note/warning/tip panel macros to <blockquote> elements.
local function convert_panel(macro_type, block)
  local body = block:match('<ac:rich%-text%-body>([%s%S]-)</ac:rich%-text%-body>') or ''
  return '<blockquote><p><strong>' .. macro_type:upper() .. ':</strong></p>\n'
      .. trim(body) .. '\n</blockquote>'
end

-- ── Main preprocessing ────────────────────────────────────────────────────────

local function preprocess(text)
  -- PlantUML round-trip: recover source from the HTML comment stashed on publish.
  -- Pattern: <!-- plantuml_src_b64: BASE64 --><p><img .../></p>
  text = text:gsub(
    '<!%-%- plantuml_src_b64: ([A-Za-z0-9+/=]+) %-%->[%s]*<p><img[^>]-/></p>',
    function(b64)
      local src = b64_decode(b64)
      if not src then return '<!-- plantuml decode failed -->' end
      return '<pre><code class="language-plantuml">' .. html_escape(src) .. '</code></pre>'
    end)

  -- Code macros — must run before the global CDATA sweep so the code body
  -- is extracted before CDATA delimiters are HTML-escaped.
  text = text:gsub(
    '<ac:structured%-macro[^>]*ac:name="code"[^>]*>[%s%S]-</ac:structured%-macro>',
    convert_code_macro)

  -- Info / note / warning / tip panel macros.
  for _, t in ipairs({ 'info', 'warning', 'note', 'tip' }) do
    text = text:gsub(
      '<ac:structured%-macro[^>]*ac:name="' .. t .. '"[^>]*>[%s%S]-</ac:structured%-macro>',
      function(block) return convert_panel(t, block) end)
  end

  -- Expand / collapsible macros — keep the rich-text body content.
  text = text:gsub(
    '<ac:structured%-macro[^>]*ac:name="expand"[^>]*>[%s%S]-</ac:structured%-macro>',
    function(block)
      local body = block:match('<ac:rich%-text%-body>([%s%S]-)</ac:rich%-text%-body>') or ''
      return trim(body)
    end)

  -- Drop any remaining structured macros (TOC, status, Jira, etc.).
  text = text:gsub('<ac:structured%-macro[^>]*>[%s%S]-</ac:structured%-macro>', '')

  -- Remaining CDATA sections (outside macros) → HTML-escaped text.
  text = text:gsub('<!%[CDATA%[([%s%S]-)%]%]>', function(c)
    return html_escape(c)
  end)

  -- Strip ac: and ri: namespace tags, keeping any inner text content.
  -- Handle both self-closing and regular open/close tags.
  text = text:gsub('<ac:[^>]-/>', '')
  text = text:gsub('</?ac:[^>]->', '')
  text = text:gsub('<ri:[^>]-/>', '')
  text = text:gsub('</?ri:[^>]->', '')

  return text
end

-- ── Entry point ───────────────────────────────────────────────────────────────

local input = io.read('*a')
io.write(preprocess(input))

File: scripts/confluence_preproc.py
#!/usr/bin/env python3
"""
confluence_preproc.py — Pre-process Confluence storage-format HTML into standard
HTML suitable for conversion to CommonMark via pandoc.

Usage:
    curl ... | python3 confluence_preproc.py | pandoc --from=html --to=commonmark

Called automatically by confluence_publish.sh --pull. Not normally invoked directly.

Transformations applied (in order):
    ac:structured-macro name="code"              → <pre><code class="language-X">
    ac:structured-macro name="info|note|warning|tip" → <blockquote>
    ac:structured-macro name="expand"            → unwrapped rich-text body
    Remaining ac:structured-macro elements       → stripped (content discarded)
    CDATA sections                               → HTML-escaped inline text
    ac:* and ri:* namespace tags                 → stripped (inner text kept)
"""

import re
import sys
import html as htmlmod
import base64


def convert_code_macro(m: re.Match) -> str:
    """Convert a Confluence code macro to an HTML <pre><code> block."""
    block = m.group(0)
    lang_m  = re.search(r'ac:name=["\']?language["\']?[^>]*>\s*(.*?)\s*</ac:parameter>', block, re.DOTALL)
    cdata_m = re.search(r'<!\[CDATA\[(.*?)\]\]>', block, re.DOTALL)
    body_m  = re.search(r'<ac:plain-text-body>(.*?)</ac:plain-text-body>', block, re.DOTALL)
    lang = lang_m.group(1).strip() if lang_m else ""
    body = cdata_m.group(1) if cdata_m else (body_m.group(1).strip() if body_m else "")
    return '<pre><code class="language-{}">{}</code></pre>'.format(lang, htmlmod.escape(body))


def convert_panel(m: re.Match) -> str:
    """Convert info/note/warning/tip macros to <blockquote> elements."""
    block = m.group(0)
    nm = re.search(r'ac:name=["\']?(\w+)["\']?', block)
    bm = re.search(r'<ac:rich-text-body>(.*?)</ac:rich-text-body>', block, re.DOTALL)
    name = nm.group(1).upper() if nm else "NOTE"
    body = bm.group(1).strip() if bm else ""
    return "<blockquote><p><strong>{}:</strong></p>\n{}\n</blockquote>".format(name, body)


def unwrap_expand(m: re.Match) -> str:
    """Unwrap expand/collapsible macros — keep the body content."""
    block = m.group(0)
    bm = re.search(r'<ac:rich-text-body>(.*?)</ac:rich-text-body>', block, re.DOTALL)
    return bm.group(1).strip() if bm else ""


def recover_plantuml(m: re.Match) -> str:
    """Recover a plantuml fenced block from the base64 source stashed on publish."""
    try:
        src = base64.b64decode(m.group(1)).decode("utf-8")
    except Exception:
        return m.group(0)  # leave untouched if decoding fails
    return '<pre><code class="language-plantuml">{}</code></pre>'.format(htmlmod.escape(src))


def preprocess(text: str) -> str:
    # PlantUML round-trip: recover source from the HTML comment stashed during publish.
    # Pattern: <!-- plantuml_src_b64: BASE64 --><p><img ...></p>  (whitespace-tolerant)
    text = re.sub(
        r'<!-- plantuml_src_b64: ([A-Za-z0-9+/=]+) -->\s*<p><img[^>]*/></p>',
        recover_plantuml, text)

    # Code macros first — extract CDATA before the global CDATA sweep
    text = re.sub(
        r'<ac:structured-macro\s+ac:name=["\']?code["\']?[^>]*>.*?</ac:structured-macro>',
        convert_code_macro, text, flags=re.DOTALL)

    # Info / note / warning / tip panels
    text = re.sub(
        r'<ac:structured-macro\s+ac:name=["\']?(?:info|warning|note|tip)["\']?[^>]*>'
        r'.*?</ac:structured-macro>',
        convert_panel, text, flags=re.DOTALL)

    # Expand / collapsible sections — unwrap the body
    text = re.sub(
        r'<ac:structured-macro\s+ac:name=["\']?expand["\']?[^>]*>.*?</ac:structured-macro>',
        unwrap_expand, text, flags=re.DOTALL)

    # Drop any remaining structured macros (TOC, status, Jira, etc.)
    text = re.sub(
        r'<ac:structured-macro[^>]*>.*?</ac:structured-macro>', "", text, flags=re.DOTALL)

    # Remaining CDATA sections (outside macros) → HTML-escaped text
    text = re.sub(
        r'<!\[CDATA\[(.*?)\]\]>',
        lambda m: htmlmod.escape(m.group(1)), text, flags=re.DOTALL)

    # Strip ac: and ri: namespace tags, keeping any inner text content
    text = re.sub(r'</?ac:[^>]+/?>', "", text)
    text = re.sub(r'</?ri:[^>]+/?>', "", text)

    return text


if __name__ == "__main__":
    print(preprocess(sys.stdin.read()))
